Change API from []byte to reflect.Value

GreenHorse opened the pull request:
Use reflect.Value high level for the package API, remove scanner package.
Add Addition and Set.
Update `README` to match changes.

GreenHorse committed:
Change api from []byte to interface{}

Instead of using a []bytes of json, we use golang interface{} and go trought the
struct using the reflect package

GreenHorse committed:
Add set and addition

GreenHorse committed:
Report inputs not applied by Addition/Set

GreenHorse committed:
Update import

GreenHorse committed:
Update Readme

GreenHorse committed:
Make test parallel and make json tag lookup safe for concurrency

minjatJ commented on README.md:
```diff
@@ -48,48 +55,65 @@ The initial goal is to support all the selectors the original jq command line su
 | syntax | value |
 | :--- | :--- |
 | .string | "a" |
-| .number| 1.23 |
```
Floats are not supported anymore ?

minjatJ commented on err.go:
```diff
@@ -0,0 +1,11 @@
+	// ErrCannotSet indicates the value is not settable, it will happend
+	// when a value is passed by copy
+	ErrCannotSet = errors.New("Cannot set value")
+	// ErrUnknownField indicates that we trie to access a field not present in
```
`we try`

minjatJ commented on op.go:
```diff
@@ -15,42 +15,231 @@
+func callChain(v reflect.Value, chainFun ...Op) (reflect.Value, error) {
+	for _, f := range chainFun {
+		var err error
+		v, err = f.Apply(v)
```
`err` should be declare at the beginning then `err != nil` condition can be merged with assignment.

GreenHorse committed:
Remove Debug, pre-compute and store reflect for Type Kind

Add a fallback for version before go1.10
Add an helper method to try to unmarshall

minjatJ commented on op.go:
```diff
@@ -15,42 +15,231 @@
+		case reflect.Map:
+			var err error
+			mapVal := in.MapIndex(reflect.ValueOf(key))
+			newMapVal := reflect.New(in.MapIndex(reflect.ValueOf(key)).Type())
```
Why not using `mapVal` here ? (`reflect.New(mapVal)...`)

GreenHorse replied:
No reason, it is even better for map of interfaces as we can obtain the concrete type for the key.

minjatJ commented on struct_json.go:
```diff
@@ -0,0 +1,36 @@
+		}
+		jsonNameToIdx.LoadOrStore(v.Type(), buf)
+	}
+	buf, _ := jsonNameToIdx.Load(v.Type())
```
Please store `v.Type()` result

minjatJ commented on op.go:
```diff
@@ -15,42 +15,231 @@
+			in.Set(reflect.AppendSlice(in, valRef))
+			return callChain(in, chainFun...)
+		case reflect.Map:
+			if v, ok := val.(json.RawMessage); ok {
```
`case reflect.Slice` has similar code. We should implement a `valueFromJSON` method

minjatJ commented on op.go:
```diff
@@ -15,42 +15,231 @@
+// Addition adds the val parameter to the provided interface{} (map/slice/struct)
+func Addition(val interface{}, chainFun ...Op) OpFunc {
+	valRef := reflect.ValueOf(val)
+
```
We should maybe precompute `valRef.Kind/Type`

minjatJ commented on op.go:
```diff
@@ -15,42 +15,231 @@
+
+// Set change the val parameter to the provided interface{} (map/slice/struct)
+func Set(val interface{}, chainFun ...Op) OpFunc {
+	valRef := reflect.ValueOf(val)
```
ditto

minjatJ commented on parse.go:
```diff
@@ -34,29 +38,189 @@ func Must(op Op, err error) Op {
-func Parse(selector string) (Op, error) {
+// TODO: move the parsing logic to another package
+func Parse(selector string, args ...interface{}) (Op, error) {
+	log.SetFlags(log.Llongfile)
```
Debug code ?

minjatJ commented on parse.go:
```diff
@@ -34,29 +38,189 @@ func Must(op Op, err error) Op {
+	return func(op ...Op) Op {
+		return Addition(arg, op...)
+	}, true
+	// return nil, false
```
dead code

minjatJ reviewed (changes requested):
Nice work, only nit. Keep in mind that reflection is quite heavy so we should precompute as much as we can

minjatJ commented on decode_go1.10.go:
```diff
@@ -0,0 +1,14 @@
+// +build go1.10
```
Good idea

minjatJ reviewed (approved):
Thanks for the changes. Nice one, LGTM
